File: v2/robotmk/parse_xml.py
import enum
from datetime import datetime
from typing import Annotated, TypeVar

import xmltodict
from pydantic import BaseModel, BeforeValidator, Field, PlainValidator


def parse_datetime(raw: str) -> datetime:
    return datetime.strptime(raw, "%Y%m%d %H:%M:%S.%f")


T = TypeVar("T")


def parse_children(raw: T | list[T]) -> list[T]:
    return raw if isinstance(raw, list) else [raw]


DateTime = Annotated[datetime, PlainValidator(parse_datetime)]


class XML(BaseModel, frozen=True):
    ...


class Outcome(enum.Enum):
    FAIL = "FAIL"
    PASS = "PASS"
    SKIP = "SKIP"
    NOT_RUN = "NOT RUN"


class Status(XML, frozen=True):
    status: Outcome = Field(alias="@status")
    starttime: DateTime = Field(alias="@starttime")
    endtime: DateTime = Field(alias="@endtime")


class Test(XML, frozen=True):
    id_: str = Field(alias="@id")
    name: str = Field(alias="@name")
    line: int = Field(alias="@line")
    status: Status


class Suite(XML, frozen=True):
    id_: str = Field(alias="@id")
    name: str = Field(alias="@name")
    suite: Annotated[list["Suite"], BeforeValidator(parse_children)] = Field(default=[])
    test: Annotated[list[Test], BeforeValidator(parse_children)] = Field(default=[])


class Generator(XML, frozen=True):
    generator: str = Field(alias="@generator")
    generated: DateTime = Field(alias="@generated")
    rpa: bool = Field(alias="@rpa")
    # TODO: Ensure schemaversions other than `4` work.  # pylint: disable=fixme
    schemaversion: int = Field(alias="@schemaversion")
    suite: Annotated[list[Suite], BeforeValidator(parse_children)] = Field(default=[])
    errors: object  # TODO: Gracefully handle errors  # pylint: disable=fixme


class Rebot(XML, frozen=True):
    robot: Generator


def parse_rebot(xml: str) -> Rebot:
    return Rebot.model_validate(xmltodict.parse(xml))

File: v2/rust/src/parse_xml.rs
use chrono::NaiveDateTime;
use serde::de::Error;
use serde::{Deserialize, Deserializer};

#[derive(Deserialize)]
#[serde(rename_all = "UPPERCASE")]
enum Outcome {
    Fail,
    Pass,
    Skip,
    #[serde(rename = "NOT RUN")]
    NotRun,
}

fn parse_datetime<'de, D>(deserializer: D) -> Result<NaiveDateTime, D::Error>
where
    D: Deserializer<'de>,
{
    let raw = Deserialize::deserialize(deserializer)?;
    NaiveDateTime::parse_from_str(raw, "%Y%m%d %H:%M:%S.%f")
        .map_err(|_| Error::custom(format!("Invalid datetime: {}", raw)))
}

#[derive(Deserialize)]
struct Rebot {
    #[serde(rename = "@generator")]
    generator: String,
    #[serde(rename = "@generated", deserialize_with = "parse_datetime")]
    generated: NaiveDateTime,
    #[serde(rename = "@rpa")]
    rpa: bool,
    #[serde(rename = "@schemaversion")]
    schemaversion: usize,
    suite: Suite,
    errors: Errors,
}

#[derive(Deserialize)]
struct Suite {
    #[serde(rename = "@id")]
    id: String,
    #[serde(rename = "@name")]
    name: String,
    #[serde(default)]
    suite: Vec<Suite>,
    #[serde(default)]
    test: Vec<Test>,
}

#[derive(Deserialize)]
struct Status {
    #[serde(rename = "@status")]
    status: Outcome,
    #[serde(rename = "@starttime", deserialize_with = "parse_datetime")]
    starttime: NaiveDateTime,
    #[serde(rename = "@endtime", deserialize_with = "parse_datetime")]
    endtime: NaiveDateTime,
}

#[derive(Deserialize)]
struct Test {
    #[serde(rename = "@id")]
    id: String,
    #[serde(rename = "@name")]
    name: String,
    #[serde(rename = "@line")]
    line: usize,
    status: Status,
}

#[derive(Deserialize)]
struct Errors {}
